### PR TITLE
Change Doc to trait to allow it to be used as a cake slice.

### PR DESCRIPTION
**Pull request by vigdorchik**
Change the implementation of retrieve() method.
Rename and privatize symbols in CommentFactoryBase to avoid clashes when cooking a cake.

### CODE REVIEW

**Comment by vigdorchik** [comment]
Review by @huitseeker.

File: src/compiler/scala/tools/nsc/interactive/Doc.scala [hunk 1]
```diff
@@ -8,29 +8,27 @@ package interactive
...
-      else
-        if ((site.isClass || site.isModule) && site.info.members.toList.contains(sym))
-          Some(LinkToMember(sym, site))
-        else
-          None
-    }
+    if (sym.isClass || sym.isModule)
+      Some(LinkToTpl(sym))
+    else
+      if ((site.isClass || site.isModule) && site.info.members.toList.contains(sym))
```
Comment (dragos) [inline src/compiler/scala/tools/nsc/interactive/Doc.scala:30]: The call to `info.members` should be done inside an `ask`. If this is the caller's responsibility, it should be documented in the scaladoc for `internalLink` (or make the method private, if it's not supposed to be called by clients).

Comment (vigdorchik) [inline src/compiler/scala/tools/nsc/interactive/Doc.scala:30]: It is called inside comment generation (i.e. retrieve()), so it's safe.
As this method implements the interface, it cannot be made private.

**Comment by vigdorchik** [comment]
The relevant commit in Scala-IDE is https://github.com/vigdorchik/scala-ide/commit/0055190eb25dc5a2e7cdffa07d5acfd5121d9b8e#L5R13

File: src/compiler/scala/tools/nsc/interactive/Doc.scala (hunk 1 above)
Comment (huitseeker) [inline src/compiler/scala/tools/nsc/interactive/Doc.scala:30]: @vigdorchik Yes. The scaladoc remark still applies, though.

Comment (vigdorchik) [inline src/compiler/scala/tools/nsc/interactive/Doc.scala:30]: On Thu, Jan 31, 2013 at 2:59 PM, François Garillot <notifications@github.com

File: src/compiler/scala/tools/nsc/interactive/Doc.scala [hunk 2]
```diff
@@ -41,19 +42,34 @@ abstract class Doc(val settings: doc.Settings) extends MemberLookupBase with Com
...
-      resp.get.left.toOption flatMap { _ =>
-        ask { () =>
-          val comment = parseAtSymbol(expandedDocComment(sym), rawDocComment(sym), pos, Some(site))
+ /**
+  * Gets the documentation either as a Comment object or as a URL.
+  * @param sym      The symbol whose documentation should be retrieved.
+  * @param site     The place where sym is observed.
+  * @param source   The source file to look for the symbol.
+  */
+  def retrieve(sym: Symbol, site: Symbol, source: SourceFile): Option[DocResult] = {
```
Comment (huitseeker) [inline src/compiler/scala/tools/nsc/interactive/Doc.scala]: I looked for usages of this one by grepping, and found it frustrating to grep for something called `retrieve`. Would you mind renaming it to something like `retrieveDocForSymbol` (or something expressive you'd like better) ?

File: src/compiler/scala/tools/nsc/interactive/Global.scala [hunk 3]
```diff
@@ -562,7 +562,7 @@ class Global(settings: Settings, _reporter: Reporter, projectName: String = "")
   }
 
   /** Parse unit and create a name index, unless this has already been done before */
-  private def parseAndEnter(unit: RichCompilationUnit): Unit =
+  private[interactive] def parseAndEnter(unit: RichCompilationUnit): Unit =
```
Comment (huitseeker) [inline src/compiler/scala/tools/nsc/interactive/Global.scala]: You don't need to make this one object-local.

**Comment by vigdorchik** [comment]
PLS REBUILD ALL

**Pull request closed without merging**